File: 프로그래머스/2/92342. 양궁대회/양궁대회.py
def solution(n, info):
    max_diff = 0
    best_result = [-1]
    
    def bt(apeach, arrows_left, lion, idx):
        nonlocal max_diff, best_result
        
        # 모든 화살을 다 쐈거나 모든 점수를 확인한 경우
        if arrows_left == 0 or idx == 11:
            # 남은 화살은 0점에 모두 쏘기
            if arrows_left > 0:
                lion[10] = arrows_left
            
            # 점수 계산
            apeach_score, lion_score = 0, 0
            for i in range(11):
                if apeach[i] == 0 and lion[i] == 0:
                    continue
                if lion[i] > apeach[i]:
                    lion_score += (10 - i)
                else:
                    apeach_score += (10 - i)
            
            diff = lion_score - apeach_score
            
            # 더 큰 차이로 이기는 경우
            if diff > max_diff:
                max_diff = diff
                best_result = lion[:]
            # 같은 차이인 경우 낮은 점수를 더 많이 맞힌 경우
            elif diff > 0 and diff == max_diff:
                if is_better(lion, best_result):
                    best_result = lion[:]
            
            # 백트래킹 준비
            if arrows_left > 0:
                lion[10] = 0
            return
        
        # 현재 점수를 이기는 경우 (어피치보다 1발 더 쏘기)
        need = apeach[idx] + 1
        if arrows_left >= need:
            lion[idx] = need
            bt(apeach, arrows_left - need, lion, idx + 1)
            lion[idx] = 0
        
        # 현재 점수를 포기하는 경우
        bt(apeach, arrows_left, lion, idx + 1)
    
    def is_better(new, old):
        # 낮은 점수를 더 많이 맞힌 것이 유리
        for i in range(10, -1, -1):
            if new[i] > old[i]:
                return True
            elif new[i] < old[i]:
                return False
        return False
    
    lion = [0] * 11
    bt(info, n, lion, 0)
    
    return best_result

File: 프로그래머스/2/92342. 양궁대회/양궁대회.java
class Solution {
    int maxDiff = 0;
    int[] answer;
    public int[] solution(int n, int[] info) {
        int[] lionRecord = new int[11];
        answer = new int[11];
        bt(n, 0, lionRecord, info);
        return maxDiff == 0 ? new int[] {-1} : answer;
    }
    
    public void bt(int leftArrow, int idx, int[] lionRecord, int[] apeachRecord){
        if (idx == 11) {
            lionRecord[10] += leftArrow;
            int diff = calcDiff(lionRecord, apeachRecord);
            if (diff > 0 && diff >= maxDiff) {
                if (diff > maxDiff || isBetter(lionRecord)) {
                    maxDiff = diff;
                    answer = lionRecord.clone();
                }
            }
            lionRecord[10] -= leftArrow;
            return;
        }
        
        int required = apeachRecord[idx] + 1;
        if (leftArrow >= required) {
            lionRecord[idx] = required;
            bt(leftArrow - required, idx + 1, lionRecord, apeachRecord);
            lionRecord[idx] = 0;
        }
        
        // 어피치가 0발이어도 라이언이 0발 쏘면 점수 못 얻으므로 로직상 안전함
        bt(leftArrow, idx + 1, lionRecord, apeachRecord);
    }
    
    // 낮은 점수 화살이 더 많은지 체크
    private boolean isBetter(int[] lionRecord) {
        for (int i = 10; i >= 0; i--) {
            if (lionRecord[i] > answer[i]) return true;
            if (lionRecord[i] < answer[i]) return false;
        }
        return false;
    }
    
    private int calcDiff(int[] lionRecord, int[] apeachRecord){
        int lion = 0;
        int apeach = 0;
        for(int i = 10; i >= 0; i--){
            if(lionRecord[i] == 0 && apeachRecord[i] == 0) continue;
            if(lionRecord[i] > apeachRecord[i]) lion += 10 - i;
            else apeach += 10 - i;            
        }
        return lion - apeach;
    }
}